### PR TITLE
Strengthened SecurityTest, removed dead PHP < 7.4 fallbacks

### PR DESCRIPTION
**Pull request by WarLikeLaux**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Tests added?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

## What does this PR do?

Strengthens `SecurityTest` coverage and removes dead code for PHP < 7.4 from `Security`.

### Before

```
Tests: 64, Assertions: 364
Methods: 32% (6/19), Lines: 48% (83/174)
```

### After

```
Tests: 113, Assertions: 461
Methods: 67% (12/18), Lines: 92% (92/100)
MSI: 78%
```

### SecurityTest (+49 tests)

- Encrypt/decrypt: invalid cipher, tampered data, wrong key, binary data, empty string, info context, all 3 cipher variants
- Password hashing: default/custom cost, validation edge cases (empty password, invalid hash format, cost boundaries)
- Key derivation (hkdf/pbkdf2), random key/string validation, compareString type checks, hashData rawHash, maskToken/unmaskToken edge cases

### Dead code removed (~75 lines)

Removed fallback paths that are always dead on PHP 7.4+:

- `hkdf()`: manual HMAC-based fallback for PHP < 7.1 (no `hash_hkdf`)
- `pbkdf2()`: manual HMAC-based fallback + `PHP_VERSION_ID >= 50500` check
- `generatePasswordHash()`: `crypt()` fallback for PHP < 5.5 (no `password_hash`)
- `validatePassword()`: `crypt()` fallback for PHP < 5.5 (no `password_verify`)
- `generateSalt()`: only used by the `crypt()` fallback above
- `compareString()`: manual XOR fallback for PHP < 5.6 (no `hash_equals`)

Yii 2.1 was never released and the next planned version is 2.2. Similar cleanup was done in #20743 and #20756.

### Uncovered lines (8)

All `throw` inside error paths that cannot be triggered in tests: `!extension_loaded('openssl')`, `openssl_encrypt/decrypt === false`, `hash_hkdf/hash_pbkdf2 === false`, `hash_hmac` failure.

### Escaped mutants

- `shouldUseLibreSSL()`: 20 mutants - cannot be tested without LibreSSL, always returns `false` on OpenSSL
- `(string)` casts in `hash_hkdf()`: equivalent - params are already strings in tests
- Visibility mutations: equivalent - `ExposedSecurity` subclass overrides visibility

### CODE REVIEW

File: tests/framework/base/SecurityTest.php [hunk 1]
```diff
@@ -1129,4 +1129,341 @@ public static function maskProvider(): array
...
+        $key = $this->security->pbkdf2('sha256', 'password', 'salt', 1000, 0);
+        $this->assertSame(32, strlen($key));
+    }
+
+    public function testValidatePasswordAcceptsAllBlowfishPrefixes(): void
+    {
+        $this->security->passwordHashCost = 4;
+        $hash = $this->security->generatePasswordHash('test');
+        $this->assertTrue($this->security->validatePassword('test', $hash));
+        $this->assertFalse($this->security->validatePassword('wrong', $hash));
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply tests/framework/base/SecurityTest.php:1376]: Fixed.

File: tests/framework/base/SecurityTest.php [hunk 2]
```diff
@@ -1129,4 +1129,341 @@ public static function maskProvider(): array
...
+        $this->expectExceptionMessage('integer given');
+        $this->security->compareString('test', 123);
+    }
+
+    public function testEncryptByPasswordInvalidCipherShowsCipherName(): void
+    {
+        $this->security->cipher = 'BOGUS-256-XYZ';
+        $this->expectException(\yii\base\InvalidConfigException::class);
+        $this->expectExceptionMessage('BOGUS-256-XYZ is not an allowed cipher');
+        $this->security->encryptByPassword('data', 'pass');
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply tests/framework/base/SecurityTest.php]: Fixed.

File: tests/framework/base/SecurityTest.php [hunk 3]
```diff
@@ -1129,4 +1129,341 @@ public static function maskProvider(): array
...
+        $this->security->compareString(123, 'test');
+    }
+
+    public function testCompareStringsWithNonStringActualShowsType(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('integer given');
+        $this->security->compareString('test', 123);
+    }
+
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply tests/framework/base/SecurityTest.php]: Fixed.

File: framework/base/Security.php [hunk 4]
```diff
@@ -584,46 +500,7 @@ public function validatePassword($password, $hash)
...
-
-        // Get a 20-byte random string
-        $rand = $this->generateRandomKey(20);
-        // Form the prefix that specifies Blowfish (bcrypt) algorithm and cost parameter.
-        $salt = sprintf('$2y$%02d$', $cost);
-        // Append the random salt data in the required base64 format.
-        $salt .= str_replace('+', '.', substr(base64_encode($rand), 0, 22));
-
-        return $salt;
+        return password_verify($password, $hash);
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply framework/base/Security.php:503]: Fixed.

File: framework/base/Security.php [hunk 5]
```diff
@@ -547,19 +475,7 @@ public function generatePasswordHash($password, $cost = null)
...
-
-        $salt = $this->generateSalt($cost);
-        $hash = crypt($password, $salt);
-        // strlen() is safe since crypt() returns only ascii
-        if (!is_string($hash) || strlen($hash) !== 60) {
-            throw new Exception('Unknown error occurred while generating hash.');
-        }
-
-        return $hash;
+        return password_hash($password, PASSWORD_DEFAULT, ['cost' => $cost]);
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply framework/base/Security.php:478]: Fixed.

**Review by terabytesoftw — APPROVED** [review]

File: framework/base/Security.php [hunk 6]
```diff
@@ -584,46 +499,7 @@ public function validatePassword($password, $hash)
...
-     * The PHP [crypt()](https://www.php.net/manual/en/function.crypt.php) built-in function
-     * requires, for the Blowfish hash algorithm, a salt string in a specific format:
-     * "$2a$", "$2x$" or "$2y$", a two digit cost parameter, "$", and 22 characters
-     * from the alphabet "./0-9A-Za-z".
-     *
-     * @param int $cost the cost parameter
-     * @return string the random salt value.
-     * @throws InvalidArgumentException if the cost parameter is out of the range of 4 to 31.
-     */
-    protected function generateSalt($cost = 13)
```
Comment (mspirkov) [inline framework/base/Security.php:612]: Removing this method may break backward compatibility, as it is `protected`. Perhaps it would be better to mark it as deprecated and remove it in version `2.2`?

Comment (WarLikeLaux) [reply framework/base/Security.php:612]: Marked as @deprecated instead of removing.

**Review by mspirkov — APPROVED** [review]

File: tests/framework/base/SecurityTest.php [hunk 7]
```diff
@@ -1129,4 +1130,311 @@ public static function maskProvider(): array
             ['Token with UTF8 character: †'],
         ];
     }
+
+    public function testShouldUseLibreSSL(): void
+    {
+        $result = $this->security->shouldUseLibreSSL();
```
Comment (samdark) [inline tests/framework/base/SecurityTest.php]: Seems this method isn't used anywhere as well. Right?

**Review by samdark — APPROVED** [review]
Overall 👍

File: tests/framework/base/SecurityTest.php (hunk 7 above)
Comment (WarLikeLaux) [reply tests/framework/base/SecurityTest.php]: Yes, it's not called anywhere in the framework. Removed the tests and marked the method as deprecated since 2.0.55 and removed in 2.2 (same as generateSalt).

File: framework/CHANGELOG.md [hunk 8]
```diff
@@ -25,6 +25,7 @@ Yii Framework 2 Change Log
 - Bug #20751: Fix `@param` annotation for `$param` parameter in `Sort::parseSortParam()` (mspirkov)
 - Enh #20756: Remove dead code for PHP < 5.6 in `SchemaBuilderTrait::json()` (WarLikeLaux)
 - Bug #20764: Fix `@return` annotation for `Model::rules()` (mspirkov)
+- Enh #20757: Remove dead code for PHP < 7.4 in `Security` (WarLikeLaux)
```
Comment (samdark) [inline framework/CHANGELOG.md]:
```suggestion
- Chg #20757: Remove dead code for PHP < 7.4 in `Security` (WarLikeLaux)
```
